### ISSUE TITLE
“Unicode character” should likely say “Unicode scalar value” in intro to lexical grammar

### ISSUE THREAD

**Issue by bathos** [whatwg/webidl#1080]
The lexical grammar is introduced with:

> The tokenizer operates on a sequence of Unicode characters [[UNICODE]](https://www.unicode.org/versions/Unicode14.0.0/).

The Unicode standard does not define “Unicode characters” as far as I can tell, so that leaves “characters,” which it does define\*, but it the definitions (plural) don’t seem compatible with how the term is used in this context.

It seems this should probably say USVs. This can be inferred sorta because the [string literal interpretation algorithm](https://webidl.spec.whatwg.org/#string-literal) appears to assume that source text consumed by `string` is already known to be exclusively USVs. I don’t know the ins-and-outs of Perl 5.5.8 regular expressions\*\*, but it seems like the grammar given for `string`, `/"[^"]*"/`, likely doesn’t preclude lone surrogates in itself, which implies that the “operates on a sequence...” statement was meant to establish USVs-only as a prior fact about the input.

---

\* It goes into considerable detail in the “Characters, not glyphs” section of [§ 2.2 Unicode Design Principles](https://www.unicode.org/versions/Unicode14.0.0/ch02.pdf) and provides a [glossary entry for “character”](http://www.unicode.org/glossary/).

\*\* Gave up on figuring this out because IIUC perl’s strings have observable encodings and character sets that impact how its regular expressions get interpreted — so it may be the case that there isn’t a single answer to the question “does `[^"]` match non-USV code points”.

**Comment by xfq** [comment]
FYI - here's some related information from other places:

In [Infra](https://infra.spec.whatwg.org/#code-points):

> Code points are sometimes referred to as characters and in certain contexts are prefixed with "0x" rather than "U+".

In [CSS](https://drafts.csswg.org/css-syntax-3/#tokenization):

> To **tokenize** a stream of code points into a stream of CSS tokens *input*, repeatedly consume a token from *input* until an &lt;EOF-token&gt; is reached, pushing each of the returned tokens into a stream.

It seems that they have not ruled out the surrogate code points, but I am not sure how the situation here is. See also [Internationalization Best Practices for Spec Developers](https://w3c.github.io/bp-i18n-specdev/#characters).

**Comment by bathos** [comment]
In CSS it's unambiguously defined in the section prior, 3.3, ["preprocessing the input stream"](https://drafts.csswg.org/css-syntax-3/#input-preprocessing):

> Replace any U+0000 NULL or surrogate code points in input with U+FFFD REPLACEMENT CHARACTER (�).

**Comment by annevk** [comment]
I suggest that we make Web IDL reference Infra for these things and not Unicode directly.

**Comment by domenic** [comment]
> It seems this should probably say USVs. This can be inferred sorta because the string literal interpretation algorithm appears to assume that source text consumed by string is already known to be exclusively USVs.

Great find. My first instinct was that it doesn't really matter since all Web IDL constructs are ASCII. But I guess this part of the spec does indeed assume USVs.

Note that unlike, say, CSS or HTML, Web IDL doesn't really have an "entry point" for parsing, and certainly not one that's web-observable. So the choice here is really a statement about valid Web IDL files, I guess? I.e. we're saying that if some Web IDL-consuming software gets a sequence of bytes which, when decoded*, contains unpaired surrogates, then the result of that software should be a parse error.

* "Decoded": not necessarily UTF-8, as we don't (and IMO shouldn't) state anywhere that `.webidl` files must be UTF-8 encoded!

**Comment by bathos** [comment]
> So the choice here is really a statement about valid Web IDL files, I guess?

Yep. It’s totally removed from any kind of observable web platform behavior. If curious, the background for why I noticed it is that I added new API surface to my parser which accepts JS string values (interpreted Web IDL source text). Previously it only read well-formed UTF-8 buffers, so USVs were a given, but now it was something I had to consider\*. At first I thought lone surrogates should probably pass through the lexer fine if appearing in in `string`, `comment`, or `other` tokens given what’s written, but then I caught the implication for USV string literal interpretation.

> \* by “had to consider,” I mean I definitely did not have to consider it. i’m the only person using the parser in question and obviously I’m not gonna pass in any unpaired surrogates. but if i don’t invent problems, who will??

**Comment by SamB** [comment]
Well, should specs be encouraged to write IDL using non-ASCII characters? I guess maybe it could be useful in comments (though keep in mind the existance of <https://trojansource.codes>/[CVE-2021-42574](https://cve.mitre.org/cgi-bin/cvename.cgi?name=CVE-2021-42574), even if malicious spec authors would normally be sabotaging our semantics, not our IDL), but surely we want string literals to use escapes rather than non-ASCII text? (What do you mean, there *are* no escape sequences?)
